### PR TITLE
Simplify AvoidUninstantiatedInternalClasses

### PR DESCRIPTION
**Pull request by Youssef1313**

### CODE REVIEW

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidUninstantiatedInternalClasses.cs [hunk 1]
```diff
@@ -302,14 +302,8 @@ private bool HasInstantiatedNestedType(INamedTypeSymbol type, IEnumerable<INamed
...
             }
 
-            // Ignore type generated for holding top level statements
-            if (type.IsTopLevelStatementsEntryPointType())
-            {
-                return true;
-            }
-
             // The type containing the assembly's entry point is OK.
-            if (ContainsEntryPoint(type, compilation))
```
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidUninstantiatedInternalClasses.cs:312]: On roslyn side, this is calculated once for the compilation and cached:

https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs#L1676-L1684

@mavasani Do you think this PR can contribute to https://github.com/dotnet/roslyn-analyzers/issues/6301

I'm not sure how much the existing code was expensive. Specifically, we could have been calling `GetMembers` for lots of types. Did the compiler already cached them before invoking the analyzer? or does the analyzer come through this code path:

https://github.com/dotnet/roslyn/blob/1100e56a04e144352af43181df077a13957e6a58/src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs#L1656-L1665

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidUninstantiatedInternalClasses.cs [hunk 2]
```diff
@@ -360,81 +357,6 @@ private bool HasInstantiatedNestedType(INamedTypeSymbol type, IEnumerable<INamed
...
-            }
-
-            var wellKnowTypeProvider = WellKnownTypeProvider.GetOrCreate(compilation);
-            var taskSymbol = wellKnowTypeProvider.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemThreadingTasksTask);
-            var genericTaskSymbol = wellKnowTypeProvider.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemThreadingTasksTask1);
-
-            // TODO: Handle the case where Compilation.Options.MainTypeName matches this type.
-            // TODO: Test: can't have type parameters.
-            // TODO: Main in nested class? If allowed, what name does it have?
-            // TODO: Test that parameter is array of int.
```
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidUninstantiatedInternalClasses.cs:381]: All these should be handled now by the compiler :)

File: src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidUninstantiatedInternalClassesTests.cs [hunk 3]
```diff
@@ -375,7 +375,7 @@ public async Task CA1812_Basic_NoDiagnostic_MainMethodIsDifferentlyCasedAsync()
                     OutputKind = OutputKind.ConsoleApplication,
                     Sources =
                     {
-@"Friend Class C
+@"Friend Class [|C|]
```
Comment (Youssef1313) [inline src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidUninstantiatedInternalClassesTests.cs:378]: This was false negative.

**Review by mavasani — APPROVED** [review]

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidUninstantiatedInternalClasses.cs (hunk 1 above)
Comment (mavasani) [reply src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidUninstantiatedInternalClasses.cs:312]: Regardless of whether or not this addresses https://github.com/dotnet/roslyn-analyzers/issues/6301, it seems to definitely be more hardened code and should likely also improve performance.

**Pull request merged**